bin/coqtop.ml

```diff
@@ -0,0 +1,123 @@
+open Stdune
+open Import
+
+let doc =
+  "Execute the Coq toplevel with the local configuration."
+
+let man =
+  [ `S "DESCRIPTION"
+  ; `P
+      {|$(b,dune coq top FILE -- ARGS) runs the Coq toplevel to process the
+          given $(b,FILE). The given arguments are completed according to the
+          local configuration. This is equivalent to running $(b,coqtop ARGS)
+          with a $(b,_CoqProject) file containing the local configurations
+          from the $(b,dune) files, but does not require maintaining a
+          $(b,_CoqProject) file.|}
+  ; `Blocks Common.help_secs
+  ]
+
+let info = Term.info "top" ~doc ~man
+
+let term =
+  let+ common = Common.term
+  and+ context =
+    Common.context_arg ~doc:{|Run the Coq toplevel in this build context.|}
+  and+ coqtop =
+    let doc = "Run the given toplevel command instead of the default." in
+    Arg.(value & opt string "coqtop" & info [ "toplevel" ] ~docv:"CMD" ~doc)
+  and+ coq_file_arg =
+    Arg.(required & pos 0 (some string) None (Arg.info [] ~docv:"COQFILE"))
+  and+ extra_args =
+    Arg.(value & pos_right 0 string [] (Arg.info [] ~docv:"ARGS"))
+  in
+  let config = Common.init common in
+  let coqtop, argv, env =
+    Scheduler.go ~common ~config (fun () ->
+        let open Fiber.O in
+        let* setup = Import.Main.setup () in
+        let* setup = Memo.Build.run setup in
+        let sctx = Import.Main.find_scontext_exn setup ~name:context in
+        let context = Dune_rules.Super_context.context sctx in
+        (* Checks on the argument file. *)
+        let source = Common.prefix_target common coq_file_arg in
+        if not (Path.is_file (Path.of_string source)) then
+          User_error.raise [ Pp.textf "cannot find file: %s" coq_file_arg ];
```

Comment:
Is this check right? What happens if the file is generated. I'm not sure if the check is even necessary by the way, the build query should include the file as a dependency which will make this check automatic.

We should also probably add a test for this case.

Reply:
I've tried running `dune coq top` on a generated source file, and that seems to work, which is kind of weird. I agree more test cases would be useful, but I'm not quite sure how to run them. Could you explain how to do that? Is there a simple way to run a single test?

Reply:
You can run this test with `./dune.exe build @coqtop`

Reply:
Does this comment still apply?